libcst/metadata/type_inference_provider.py

```diff
@@ -0,0 +1,62 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+from typing import Dict, List, Optional, Tuple
+
+from mypy_extensions import TypedDict
+
+import libcst as cst
+from libcst import BatchableMetadataProvider
+from libcst.metadata import PositionProvider
+
+
+class Position(TypedDict):
+    line: int
+    column: int
+
+
+class Location(TypedDict):
+    path: str
+    start: Position
+    stop: Position
+
+
+class InferredType(TypedDict):
+    location: Location
+    annotation: str
+
+
+class TypeInferenceProvider(BatchableMetadataProvider[str]):
+    METADATA_DEPENDENCIES = (PositionProvider,)
+    is_cache_required = True
+
+    def __init__(self, cache: List[InferredType]) -> None:
+        super().__init__(cache)
+        lookup: Dict[Tuple[int, int, int, int], str] = {}
```

Comment:
What's the reason why  we're not using a datatype here for the key instead?  It might makes things more readable in the long term

Reply:
Change to use existing data class `CodeRange`.